tools/gbcpal.c

```diff
@@ -0,0 +1,156 @@
+#define PROGRAM_NAME "gbcpal"
+#define USAGE_OPTS "[-h|--help] [-r|--reverse] out.gbcpal [in.gbcpal...]"
+
+#include "common.h"
+
+bool reverse;
+
+void parse_args(int argc, char *argv[]) {
+	struct option long_options[] = {
+		{"reverse", no_argument, 0, 'r'},
+		{"help", no_argument, 0, 'h'},
+		{0}
+	};
+	for (int opt; (opt = getopt_long(argc, argv, "rh", long_options)) != -1;) {
+		switch (opt) {
+		case 'r':
+			reverse = true;
+			break;
+		case 'h':
+			usage_exit(0);
+			break;
+		default:
+			usage_exit(1);
+		}
+	}
+}
+
+struct Color {
+	uint8_t r, g, b;
+};
+
+uint16_t pack_color(struct Color color) {
+	return (color.b << 10) | (color.g << 5) | color.r;
+}
+
+struct Color unpack_color(uint8_t lo, uint8_t hi) {
+	return (struct Color){
+		.r = lo & 0x1f,
+		.g = ((hi << 3) & 0x18) | ((lo >> 5) & 0x07),
+		.b = (hi >> 2) & 0x1f,
+	};
+}
+
+double luminance(struct Color color) {
+	return 0.299 * color.r * color.r + 0.587 * color.g * color.g + 0.114 * color.b * color.b;
+}
+
+int compare_colors(const void *color1, const void *color2) {
+	double lum1 = luminance(*(const struct Color *)color1);
+	double lum2 = luminance(*(const struct Color *)color2);
+	return lum1 < lum2 ? 1 : lum1 > lum2 ? -1 : 0;
```

Comment:
They're doubles so that wouldn't quite work (and let's *not* have it depend on how a difference of +0.123 or -0.456 get converted to int).

Also even if you are comparing two ints, in the general case, using subtraction to do so in C is [not foolproof](https://flak.tedunangst.com/post/subtraction-is-not-comparison).

Reply:
I've updated this to be terser.